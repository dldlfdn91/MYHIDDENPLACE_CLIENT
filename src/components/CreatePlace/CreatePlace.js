import React, { Component } from "react";
import SimpleMDE from "react-simplemde-editor";
import PlacesAutocomplete, {
  geocodeByAddress,
  getLatLng
} from "react-places-autocomplete";
import "./CreatePlace.scss";
import "easymde/dist/easymde.min.css";

class CreatePlace extends Component {
  constructor(props) {
    super(props);

    this.state = {
      placeInfoGroup: {
        title: "",
        description: "",
        address: "",
        latlng: "",
        imgfile: ""
      },
      tags: [],
      inputTag: "",
      imagePreviewUrl: ""
    };

    this.handleTitleChange = this.handleTitleChange.bind(this);
    this.handleImageChange = this.handleImageChange.bind(this);
    this.handleDescriptionChange = this.handleDescriptionChange.bind(this);
    this.handleAddressChange = this.handleAddressChange.bind(this);
    this.handleSelect = this.handleSelect.bind(this);

    this.handleInputTagChange = this.handleInputTagChange.bind(this);
    this.handleTagKeyDown = this.handleTagKeyDown.bind(this);
    this.handleTagRemove = this.handleTagRemove.bind(this);

    this.submitPlaceForm = this.submitPlaceForm.bind(this);
  }

  handleTitleChange = event => {
    const placeInfoGroup = this.state.placeInfoGroup;
    placeInfoGroup[event.target.name] = event.target.value;

    this.setState({
      placeInfoGroup
    });
  };

  handleImageChange = event => {
    event.preventDefault();

    const reader = new FileReader();
    const file = event.target.files[0];
    const placeInfoGroup = this.state.placeInfoGroup;
    placeInfoGroup[event.target.name] = file;

    reader.onloadend = () => {
      this.setState({
        placeInfoGroup,
        imagePreviewUrl: reader.result
      });
    };

    reader.readAsDataURL(file);
  };

  handleDescriptionChange = description => {
    const placeInfoGroup = this.state.placeInfoGroup;
    placeInfoGroup["description"] = description;

    this.setState({
      placeInfoGroup
    });
  };

  handleAddressChange = address => {
    const placeInfoGroup = this.state.placeInfoGroup;
    placeInfoGroup["address"] = address;

    this.setState({
      placeInfoGroup
    });
  };

  handleSelect = address => {
    geocodeByAddress(address)
      .then(results => getLatLng(results[0]))
      .then(latLng => {
        const placeInfoGroup = this.state.placeInfoGroup;
        placeInfoGroup["latlng"] = latLng;
        this.setState({
          placeInfoGroup
        });
      })
      .catch(error => console.error("Error", error));
  };

  handleInputTagChange(event) {
    this.setState({ inputTag: event.target.value });
  }

  handleTagKeyDown(event) {
    if (event.keyCode === 13) {
      const { value } = event.target;
      event.preventDefault();

      this.setState(state => ({
        tags: [...state.tags, value],
        inputTag: ""
      }));
    }

    if (
      this.state.tags.length &&
      event.keyCode === 8 &&
      !this.state.inputTag.length
    ) {
      this.setState(state => ({
        tags: state.tags.slice(0, state.tags.length - 1)
      }));
    }
  }

  handleTagRemove(index) {
    return () => {
      this.setState(state => ({
        tags: state.tags.filter((tag, i) => i !== index)
      }));
    };
  }

  submitPlaceForm = event => {
    event.preventDefault();

    const formData = new FormData();
    formData.append("imgfile", this.state.placeInfoGroup.imgfile);

    const placeInfo = this.state.placeInfoGroup;
    const tags = this.state.tags;
    this.props.onSubmit(formData, tags, placeInfo);

    const placeInfoGroup = {};
    placeInfoGroup["title"] = "";
    placeInfoGroup["description"] = "";
    placeInfoGroup["address"] = "";
    placeInfoGroup["latlng"] = "";
    placeInfoGroup["imgfile"] = "";

    this.setState({
      placeInfoGroup,
      tags: [],
      imagePreviewUrl: ""
    });
  };

  render() {
    let { imagePreviewUrl } = this.state;
    let imgfilePreview = null;
    if (imagePreviewUrl) {
      imgfilePreview = (
        <img className="previewImg" src={imagePreviewUrl} alt="Selected file" />
      );
    } else {
      imgfilePreview = (
        <div className="previewText">Please select an Image for Preview</div>
      );
    }

    return (
      <div className="CreatePlace">
        <div className="place-container">
          <div className="imgPreview">{imgfilePreview}</div>
          <form onSubmit={event => this.submitPlaceForm(event)}>
            <input
              type="file"
              name="imgfile"
              onChange={event => this.handleImageChange(event)}
            />

            <div className="place-title-wrapper">
              <label>제목</label>
              <input
                type="text"
                className="place-title"
                name="title"
                placeholder="제목을 입력하세요."
                onChange={event => this.handleTitleChange(event)}
              />
            </div>
            <SimpleMDE
              label="내용"
              value={this.state.placeInfoGroup.description}
              onChange={this.handleDescriptionChange}
              options={{
                autofocus: true,
                spellChecker: false
              }}
            />

            <PlacesAutocomplete
              value={this.state.placeInfoGroup.address}
              onChange={this.handleAddressChange}
              onSelect={this.handleSelect}
            >
              {({
                getInputProps,
                suggestions,
                getSuggestionItemProps,
                loading
              }) => (
                <div>
                  <input
                    {...getInputProps({
                      placeholder: "장소를 입력하세요.",
                      className: "location-search-input"
                    })}
                  />
                  <div className="autocomplete-dropdown-container">
                    {loading && <div>Loading...</div>}
                    {suggestions.map(suggestion => {
                      const className = suggestion.active
                        ? "suggestion-item--active"
                        : "suggestion-item";
                      // inline style for demonstration purpose
                      const style = suggestion.active
                        ? { backgroundColor: "#fafafa", cursor: "pointer" }
                        : { backgroundColor: "#ffffff", cursor: "pointer" };
                      return (
                        <div
                          {...getSuggestionItemProps(suggestion, {
                            className,
                            style
                          })}
                        >
                          <span>{suggestion.description}</span>
                        </div>
                      );
                    })}
                  </div>
                </div>
              )}
            </PlacesAutocomplete>

            <div className="tag-container">
              <ul className="tag-list">
                {this.state.tags.map((tag, index) => (
                  <li
                    key={index}
                    className="tag"
                    onClick={this.handleTagRemove(index)}
                  >
                    {tag}
                    <span>(x)</span>
                  </li>
                ))}
                <input
                  value={this.state.inputTag}
                  onChange={this.handleInputTagChange}
                  onKeyDown={this.handleTagKeyDown}
                  placeholder="태그를 입력하세요."
                />
              </ul>
            </div>

            <button type="submit" className="submit-place-button">
              제출
            </button>
          </form>
        </div>
      </div>
    );
  }
}

export default CreatePlace;
