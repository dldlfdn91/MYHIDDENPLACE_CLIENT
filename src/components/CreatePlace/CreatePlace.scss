$mainColor: #1b5ac2;
$subColor: #0c2236;
$type1: #afb1be;
$type2: #e9ecef;
$font: "Noto Sans KR", sans-serif;

.CreatePlace {
  width: 48%;
  margin: 0 auto;
  padding-top: 8%;
  font-family: $font;
  .place-container {
    .place-title-wrapper {
      margin: 5% 0;
      text-align: left;
      label {
        display: none;
      }
      .place-title {
        padding: 10px 0;
        font-size: 35px;
        border: none;
        outline: none;
      }
    }
    .imgPreview {
      position: relative;
      width: 100%;
      height: 450px;
      margin: 0 auto;
      margin-bottom: 5%;
      background-color: $type2;
      border: 1px solid $type2;
      border-radius: 5px;
      flex: 1;
      justify-content: center;
      align-items: center;
      overflow: hidden;
      .previewBox {
        width: 100%;
        height: 100%;
        color: $type1;
        .previewImg {
          width: 100%;
        }
        .previewText {
          margin-top: 25%;
        }
      }
      .select-file-box {
        display: none;
      }
      &:hover {
        .img-select-box {
          background: $type1;
          opacity: 0.5;
          width: 100%;
          height: 100%;
          .previewText {
            display: none;
          }
          .select-file-box {
            position: absolute;
            display: flex;
            justify-content: center;
            align-items: center;
            top: 50%;
            left: 50%;
            transform: translate(-50%, -50%);
            label {
              display: inline;
              padding: 10px 20px;
              font-size: 15px;
              background-color: #ffffff;
              border: 1px solid $type1;
              border-radius: 5px;
              color: $subColor;
              cursor: pointer;
            }
            input {
              position: absolute;
              width: 1px;
              height: 1px;
              padding: 0;
              margin: -1px;
              overflow: hidden;
              border: none;
            }
          }
        }
      }
    }

    .place-search{
      width: 100%;
      margin: 10px 0 5px 0;
      padding: 10px;
      outline: none;
      font-size: 15px;
      text-align: center;
      border: 1px solid $type2;
      border-radius: 5px;
    }

    .tag-container {
      margin: 10px 0 5px 0;
      .tag-list {
        border: 1px solid $type2;
        border-radius: 5px;
        .tag {
          display: inline-flex;
          margin-right: 10px;
          padding-left: 5px;
          color: $mainColor;
          border: 1px solid $mainColor;
          border-radius: 5px;
          cursor: pointer;
          .delete-tag-button {
            margin: 0 5px 2px 10px;
            margin-left: 10px;
            margin-bottom: 2px;
            font-size: 16px;
          }
        }
        input {
          padding: 10px;
          font-size: 15px;
          border: none;
          outline: none;
        }
      }
    }
    .submit-place-button {
      width: 100%;
      padding: 10px;
      font-size: 15px;
      border: 1px solid $type2;
      border-radius: 5px;
      color: $subColor;
      cursor: pointer;
      &:hover {
        color: #ffffff;
        background-color: $mainColor;
        transition: all ease 0.8s;
      }
    }
  }
}
