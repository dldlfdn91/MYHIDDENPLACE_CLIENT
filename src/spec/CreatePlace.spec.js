import React from "react";
import { shallow } from "enzyme";
import CreatePlace from "../components/CreatePlace/CreatePlace";

describe("<CreatePlace />", () => {
  let wrapper;
  const props = {
    onSubmit: jest.fn()
  };

  beforeEach(() => {
    wrapper = shallow(<CreatePlace {...props} />);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("matches snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("must have place property", function() {
    expect(
      wrapper.contains(
        <button type="submit" className="submit-place-button">
          제출
        </button>
      )
    ).toBe(true);
  });
});
