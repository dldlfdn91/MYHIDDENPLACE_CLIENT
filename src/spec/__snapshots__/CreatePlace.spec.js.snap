// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreatePlace /> matches snapshot 1`] = `ShallowWrapper {}`;
